Sources/SwiftFusion/Nonlinear/NonlinearFactorGraph.swift

```diff
@@ -0,0 +1,114 @@
+import TensorFlow
+
+public struct NonlinearFactorGraph {
+  var factor1s: [Factor1] = []
+  var factor2s: [Factor2] = []
+
+  var factorCount: Int { factor1s.count + factor2s.count }
+
+  public init() {}
+
+  public mutating func add(
+    _ value1Index: Int, _ error: @escaping @differentiable (Pose2) -> Pose2, _ noise: NoiseModel
+  ) {
+    factor1s.append(Factor1(value1Index: value1Index, error: error, noise: noise))
+  }
+
+  public mutating func add(
+    _ value1Index: Int, _ value2Index: Int,
+    _ error: @escaping @differentiable (Pose2, Pose2) -> Pose2,
+    _ noise: NoiseModel
+  ) {
+    factor2s.append(Factor2(
+      value1Index: value1Index, value2Index: value2Index, error: error, noise: noise))
+  }
+
+  public func loss(at point: [Pose2]) -> Double {
+    factor1s.map { $0.noise.loss($0.error(point[$0.value1Index])) }.reduce(0, +) +
```

Comment:
just one set of factors, but this looks good, if a bit concise :-) I think it might look better with explicit arguments in the lambda. "Typing is not the bottleneck"